% Definitions
\newcommand\wtfgenes{WTFgenes}

% Abstract
\structabs{
A common technique for interpreting experimentally-identified lists of genes
is to look for enrichment of genes associated with particular ontology terms.
The most common test uses the hypergeometric distribution;
more recently, a model-based test was proposed.
These approaches must typically be run using downloaded software, or on a server.
}{
We develop a collapsed likelihood for model-based gene set analysis and present \wtfgenes, an implementation of both hypergeometric and model-based
approaches, that can be published as a static
site with computation run in JavaScript on the user's web browser client.
Apart from hosting files, zero server resources are required: the site can (for example) be served
directly from Amazon S3 or GitHub Pages.
A C++11 implementation yielding identical results runs roughly twice as fast as the JavaScript version.
}{
\wtfgenes\ is available from \url{https://github.com/evoldoers/wtfgenes} under the BSD3 license.
A demonstration for the Gene Ontology is usable at \url{https://evoldoers.github.io/wtfgo}.
}{
Ian Holmes {\tt ihholmes+wtfgenes@gmail.com}.
}{
None.
}

\section*{Introduction}

Term Enrichment Analysis (TEA) is a common technique for finding functional patterns,
specifically over-represented ontology terms, in a set of experimentally identified genes \citep{pmid15297299}.
The most common approach,
which we refer to as {\em Frequentist TEA},
is a one-tailed Fisher's Exact Test (based on the hypergeometric distribution,
which models the number of term-associations if the gene set was chosen by chance),
with a suitable correction for multiple hypothesis testing.
Frequentist TEA has been implemented many times on various platforms \citep{pmid12431279,pmid11829497,pmid12702209,pmid15297299,pmid18511468,pmid22543366,pmid23868073,pmid23586463}.

A model-based alternative to Frequentist TEA, which more directly addresses some of the multiple testing issues
(for example, by modeling the ways that an observed gene list can be broken down into complementary gene sets),
is {\em Bayesian TEA}.
In contrast to Frequentist TEA, which just rejects a null hypothesis that genes are chosen by chance,
the Bayesian TEA explicitly models the alternative hypothesis that the gene set was generated from a few random ontology terms.
This approach was introduced by \cite{pmid18676451} and further developed by \cite{pmid20172960},
who implemented model-based testing in Java and R \citep{pmid21561920}.
However, the model-based approach remains significantly less well-explored than frequentist approaches.

\begin{figure}
\includegraphics[width=\columnwidth]{model}
\caption{
  \label{fig:model}
  Model-based explanation of observed genes ($O_i$) using ontology terms ($T_j$), following \cite{pmid20172960}.
  Other variables and hyperparameters are defined in the text.
  Circular nodes indicate continuous-valued variables or hyperparameters;
  square nodes indicate discrete-valued (boolean) variables.
  Dashed lines indicate deterministic relationships;
  shaded nodes indicate observations.
  Plates (rounded rectangles) indicate replicated subgraph structures.
}
\end{figure}

The graphical model underpinning Bayesian TEA is sketched in Figure~\ref{fig:model}.
For each of the $m$ terms there
is a boolean random variable
$T_j$ (``term $j$ is activated'').
For each of the $n$ genes there is a directly-observed boolean random variable
$O_i$ (``gene $i$ is observed in the gene set''),
and one deterministic boolean variable
$H_i$ (``gene $i$ is activated'')
defined by $H_i = 1 - \prod_{j \in G_i} (1 - T_j)$
where $G_i$ is the set of terms associated with gene $i$
(including directly annotated terms, as well as ancestral terms implied by transitive closure of the directly annotated terms).
The probability parameters are $\pi$ (term activation), $\alpha$ (false positive) and $\beta$ (false negative),
and the respective hyperparameters are ${\bf p}=(p_0,p_1)$, ${\bf a}=(a_0,a_1)$ and ${\bf b}=(b_0,b_1)$.
The model is
\begin{eqnarray*}
P(T_j=1|\pi) & = & \pi \\
P(O_i=1|H_i=0,\alpha) & = & \alpha \\
P(O_i=1|H_i=1,\beta) & = & 1-\beta
\end{eqnarray*}
with
$\pi \sim \mbox{Beta}({\bf p})$,
$\alpha \sim \mbox{Beta}({\bf a})$ and
$\beta \sim \mbox{Beta}({\bf b})$.
The model of \cite{pmid20172960} is similar but used an
{\em ad hoc} discretized prior for $\pi$, $\alpha$ and $\beta$.

Most Bayesian and Frequentist TEA implementations are designed for desktop use.
Several Frequentist TEA implementations are designed for the web, such as
DAVID-WS \citep{pmid22543366} and
Enrichr \citep{pmid23586463} % \citep{pmid23586463,pmid25971742,pmid27141961}
which has a rich dynamic web front-end.
However, web-facing Frequentist TEA implementations generally require a server-hosted back end that executes code.
Further, there are no web-based Bayesian TEA implementations.

\section*{Results}

In developing our Bayesian TEA sampler,
we introduce a collapsed version of the model in Figure~\ref{fig:model} by integrating out the probability parameters.
Let $c_p = \sum_j^m T_j$ count the number of activated terms,
$c_g = \sum_i^n H_i$ the activated genes,
$c_a = \sum_i^n O_i(1-H_i)$ the false positives and
$c_b = \sum_i^n O_i H_i$ the false negatives.
Then
\[
P({\bf T},{\bf O}|{\bf a},{\bf b},{\bf p}) =
Z(c_p;m,{\bf p})
Z(c_a;n-c_g,{\bf a})
Z(c_b;c_g,{\bf b})
\]
where
\[
Z(k;N,{\bf A}) =
\frac{B(N-k+A_0,k+A_1)}{B(A_0,A_1)}
\]
is the beta-Bernoulli distribution for $k$ ordered successes in $N$ trials with hyperparameters ${\bf A}=(A_0,A_1)$,
using the beta function
\[
B(x,y) = \int_0^1 t^{x-1} (1-t)^{y-1} dt = \frac{\Gamma(x)\Gamma(y)}{\Gamma(x+y)}
\]
Integrating out probability parameters improves sampling efficiency
and allows for higher-dimensional models where, for example, we observe multiple gene sets
and give each term its own probability $\pi_j$
or each gene its own error rates $(\alpha_i, \beta_i)$.
Our implementation by default uses uninformative priors with hyperparameters ${\bf a}={\bf b}={\bf p}=(1,1)$
but this can be overridden by the user.

The MCMC sampler uses a Metropolis-Hastings kernel \citep{GilksEtal96}.
Each proposed move perturbs some subset of the term variables.
The moves include {\em flip}, where a single term is toggled;
{\em step}, where any activated term and any one of its unactivated ancestors or descendants are toggled;
{\em jump}, where any activated term and any unactivated term are toggled; and
{\em randomize}, where all term variables are uniformly randomized.
The relative rates of these moves can be set by the user.

The sampler of \cite{pmid20172960} implemented only the {\em flip} move.
To test the relative efficacy of the newly-introduced moves we measured the autocorrelation of the term variables
for one of the GO Project's test sets, containing 17 {\em S.cerevisiae}
mating genes\footnote{Gene IDs: STE2, STE3, STE5, GPA1, SST2, STE11, STE50, STE20, STE4, STE18, FUS3, KSS1, PTP2, MSG5, DIG1, DIG2, STE12.}.
The results, shown in Figure~\ref{fig:termauto}, led us to set the MCMC defaults such that
the {\em flip}, {\em step}, and {\em jump} moves are equiprobable,
while {\em randomize} is disabled.

\begin{figure}
\includegraphics[width=\columnwidth]{termAutoCorrelation}
\caption{
  \label{fig:termauto}
  Autocorrelation of term variables, as a function of the number of MCMC samples, for several MCMC kernels on a set of 17 {\em S.cerevisiae} mating genes.
  A rapidly-decaying curve indicates an efficiently-mixing kernel.
  The kernel incorporating {\em flip}, {\em step} and {\em jump} moves (defined in the text) mixes most efficiently.
}
\end{figure}

\begin{figure}
\includegraphics[width=\columnwidth]{allsim.pdf}
\caption{
  \label{fig:simresults}
  ROC curves for Frequentist and Bayesian TEA. The axes are scaled per term.
  There are 5,919 ontology terms annotated to {\em S.cerevisiae} genes, so (for example)
  a false discovery rate of 0.001 corresponds to about 6 falsely reported terms.
}
\end{figure}

We have implemented both Frequentist TEA (with Bonferroni correction) and Bayesian TEA (as described above), in both C++11 and JavaScript.
The JavaScript version can be run as a command-line tool using node, or via a web interface in a browser, and includes extensive unit tests.
The two implementations use the same random number generator and yield numerically identical results.
The C++ version is about twice as fast:
a benchmark of Bayesian TEA on a late-2014 iMac (4GHz Intel Core i7),
using the abovementioned 17 yeast mating genes and the relevant subset of 518 GO terms, run for 1,000 samples per term,
took 37.6 seconds of user time for the C++ implementation and 79.8 seconds in JavaScript.

By contrast, the Frequentist TEA approach is almost instant.
However, its weaker statistical power is apparent from Figure~\ref{fig:simresults},
which compares the recall {\em vs} specificity of Bayesian and Frequentist methods
on simulated datasets. For values of $N$ from 1 to 4, we sampled $N$ terms
from the {\em S.cerevisiae} subset of the Gene Ontology,
and generated a corresponding set of yeast genes with false positive rate 0.1\% and false negative rate 1\%.
The MCMC sampler was run for 100 iterations per term, and this experiment was repeated 100 times.
The model-based approach has vastly superior recall to the Fisher exact test, and the difference
grows with the number of terms.

Our JavaScript software, when used as a web application,
offers a ``quick report'' view using Frequentist TEA.
For the slower-running but more powerful Bayesian TEA, the software plots the log-likelihood during an MCMC sampling run, for visual feedback.
The repository includes setup scripts allowing the tool to be deployed as a ``static site'',
i.e. consisting only of static files (HTML, CSS, JSON, and JavaScript) that can be hosted via a minimal web server with no need for dynamic code execution.
This has considerable advantages: static web hosting is generally much cheaper, and far more secure, than running server-hosted web applications.

An example wtfgenes static site, configured for the GO-basic ontology and GO-annotated genomes from the Gene Ontology website,
can be found at \url{https://evoldoers.github.io/wtfgo}.

\section*{Discussion}

JavaScript genome browsers such as JBrowse \citep{pmid27072794}
represent a broader web trend of producing static sites where possible, for reasons of security and performance.
We have implemented such a static site generator for ontological term enrichment analysis of gene sets that offers both Bayesian and frequentist tests.
In contrast with existing web services for Frequentist TEA, such as DAVID-WS or Enrichr,
it requires no server resources and allows comparison of Bayesian and Frequentist approaches.

Model-based TEA is versatile: it can readily be extended
to allow for datasets that are structured
temporally \citep{pmid26111374},
spatially \citep{pmid26877824},
or by genomic region \citep{pmid20436461};
to use domain-specific biological knowledge \citep{pmid24675718};
or to incorporate additional lines of evidence such as quantitative data \citep{pmid21599902}.
We hope our development of a collapsed likelihood, and evaluation of different MCMC kernels, will assist these efforts.

Coincidentally, Fisher's Exact Test---which we call Frequentist TEA---was originally motivated by a blind tea-tasting challenge \citep{Fisher1935}.

\section*{Funding}

IHH was partially supported by NHGRI grant HG004483.
CJM was partially supported by Office of the Director R24-OD011883 and by the Director, Office of Science, Office of Basic Energy Sciences, of the US Department of Energy under Contract No. DE-AC02-05CH11231.

\bibliographystyle{natbib}
%\bibliographystyle{bioinformatics}
%\bibliographystyle{achemnat}
%\bibliographystyle{plainnat}
%\bibliographystyle{abbrv}
%
%\bibliographystyle{plain}
%
%\bibliography{Document}


\bibliography{references}
